node-api/backend/validator.go

```diff
@@ -21,80 +21,180 @@
 package backend
 
 import (
+	"bytes"
+	"slices"
+	"strconv"
+
+	"github.com/berachain/beacon-kit/consensus-types/types"
+	"github.com/berachain/beacon-kit/errors"
 	"github.com/berachain/beacon-kit/node-api/backend/utils"
 	beacontypes "github.com/berachain/beacon-kit/node-api/handlers/beacon/types"
+	"github.com/berachain/beacon-kit/primitives/crypto"
 	"github.com/berachain/beacon-kit/primitives/math"
+	statedb "github.com/berachain/beacon-kit/state-transition/core/state"
 )
 
+// FilteredValidators will grab all of the validators from the state at the
+// given slot. It will then filter them by the provided ids and statuses.
+func (b Backend) FilteredValidators(
+	slot math.Slot, ids []string, statuses []string,
+) ([]*beacontypes.ValidatorData, error) {
+	st, _, err := b.stateFromSlot(slot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get state from slot %d", slot)
+	}
+
+	validators, err := st.GetValidators()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validators")
+	}
+
+	validatorData := make([]*beacontypes.ValidatorData, 0, len(validators))
+	for _, validator := range validators {
+		data, errInProcess := b.processValidator(st, validator, slot, ids, statuses)
+		if errInProcess != nil {
+			return nil, errors.Wrapf(errInProcess, "failed to process validator")
+		}
+		if data != nil {
+			validatorData = append(validatorData, data)
+		}
+	}
+	return validatorData, nil
+}
+
+func (b Backend) processValidator(
+	st *statedb.StateDB,
+	validator *types.Validator,
+	slot math.Slot,
+	ids []string,
+	statuses []string,
```

Comment:
I may be wrong but one change I see wrt to previous implementation is that with this schema we allow to filter for both ids and statuses at the same time?
Not sure if this is a problem (I guess not), just highlighting a diff in the implementation.
cc @berachain/core

Reply:
Yeah, this is as per the beacon API spec https://ethereum.github.io/beacon-APIs/#/Beacon/postStateValidators. Quoting `The lists of validator IDs and statuses to filter on`.

Reply:
Lgtm